provider/factorial/src/main.rs

```diff
@@ -0,0 +1,75 @@
+//! Factorial example capability provider
+//!
+//!
+#[allow(unused_imports)]
+use log::{debug, error, info, trace, warn};
+use wasmbus_rpc::provider::prelude::*;
+use wasmcloud_interface_factorial::{Factorial, FactorialReceiver};
+
+// main (via provider_main) initializes the threaded tokio executor,
+// listens to lattice rpcs, handles actor links,
+// and returns only when it receives a shutdown message
+//
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    provider_main(FactorialProvider::default())?;
+
+    eprintln!("Factorial provider exiting");
+    Ok(())
+}
+
+/// Factorial capability provider implementation
+#[derive(Default, Clone)]
+struct FactorialProvider {}
+
+/// use default implementations of provider message handlers
+impl ProviderDispatch for FactorialProvider {}
+impl FactorialReceiver for FactorialProvider {}
+impl ProviderHandler for FactorialProvider {}
+
+/// Handle Factorial methods
+#[async_trait]
+impl Factorial for FactorialProvider {
+    /// accepts a number and calculates its factorial
+    async fn calculate(&self, _ctx: &Context, req: &u32) -> RpcResult<u64> {
+        Ok(n_factorial(*req))
+    }
+}
+
+/// calculate n factorial
+fn n_factorial(n: u32) -> u64 {
+    match n {
+        0 => 1,
+        1 => 1,
+        _ => {
+            let mut result = 1u64;
+            // add 1 because rust ranges exclude upper bound
+            for v in 2..(n + 1) {
+                result *= v as u64;
+            }
+            result
+        }
+    }
+}
+
+/// Handle incoming rpc messages and dispatch to applicable trait handler.
```

Comment:
Is this generated code or code that a human has to write? If the latter, we definitely need to figure out a better way. If the former, then we should move this _out_ of the `main.rs` file and somewhere else as we shouldn't ever mix modules that contain codegen output with modules that contain developer-edited code. 

If this is "boilerplate" that the developer doesn't need to worry about then that's even more argument for this being moved out of the main file and into something less conspicuous.

Reply:
The code from line 54 down is boilerplate that will be implemented by a derive macro. The implementation of the derive macro was on hold pending a decision about whether method names would have the service prefix. Now that that's resolved, there are no more blockers (except time) for completing the derive macro, which is in a [separate gh issue](https://github.com/wasmCloud/weld/issues/27) The need for a macro affects all of the example capability providers (in rust) so it doesn't need to hold up merging this PR.  Thanks.